src/client/mod.rs

```diff
@@ -632,65 +636,33 @@ impl<C: Clone, B> Clone for Config<C, B> {
 // ===== impl Exec =====
 
 #[derive(Clone)]
-enum Exec {
-    Default,
-    Executor(Arc<Executor<Background> + Send + Sync>),
-}
-
+struct Exec(Box<CloneBoxedExecutor>);
 
 impl Exec {
-    pub(crate) fn new<E: Executor<Background> + Send + Sync + 'static>(executor: E) -> Exec {
-        Exec::Executor(Arc::new(executor))
+    pub(crate) fn new<E: Executor + Clone + Send + Sync + 'static>(executor: E) -> Exec {
+        Exec(Box::new(executor))
     }
 
-    fn execute<F>(&self, fut: F) -> io::Result<()>
+    fn execute<F>(&mut self, fut: F) -> io::Result<()>
     where
-        F: Future<Item=(), Error=()> + Send + 'static,
+        F: Future<Item=(), Error=Never> + Send + 'static,
     {
-        match *self {
-            Exec::Default => spawn(fut),
-            Exec::Executor(ref e) => {
-                e.execute(bg(Box::new(fut)))
-                    .map_err(|err| {
-                        debug!("executor error: {:?}", err.kind());
-                        io::Error::new(
-                            io::ErrorKind::Other,
-                            "executor error",
-                        )
-                    })?
-            },
-        }
-        Ok(())
+        self.0.spawn(Box::new(fut)).map_err(|err| {
+            debug!("executor error: {:?}", err);
+            io::Error::new(
+                io::ErrorKind::Other,
+                "executor error",
+            )
+        })
     }
 }
 
-// ===== impl Background =====
-
-// The types inside this module are not exported out of the crate,
-// so they are in essence un-nameable.
-mod background {
-    use futures::{Future, Poll};
-
-    // This is basically `impl Future`, since the type is un-nameable,
-    // and only implementeds `Future`.
-    #[allow(missing_debug_implementations)]
-    pub struct Background {
-        inner: Box<Future<Item=(), Error=()> + Send>,
-    }
-
-    pub fn bg(fut: Box<Future<Item=(), Error=()> + Send>) -> Background {
-        Background {
-            inner: fut,
-        }
-    }
-
-    impl Future for Background {
-        type Item = ();
-        type Error = ();
-
-        fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-            self.inner.poll()
-        }
+impl Default for Exec {
+    fn default() -> Exec {
```

Comment:
I'm not sure about that, actually. I know that this is how `CpuPool` used to work, but I've recently worked with `ThreadPool` and I'm relatively sure that it will multiplex multiple futures onto a single thread.

Effectively each thread is a while loop that reads tasks off a channel and polls them once, then repeats. The waker for each task will cause it to be re-submitted to the channel so it'll get polled again.

Can you take a look at https://github.com/rust-lang-nursery/futures-rs/blob/master/futures-executor/src/thread_pool.rs#L124 and let me know if your understanding matches mine?

Reply:
Yeah, it is pretty unfortunate. I guess the way to think about it is `cx.spawn` should be used for short-lived futures (or futures that'll either likely complete within the lifetime of the task, or that are not essential to complete).

For long lives futures (such as pool timers and connection management), I believe we need to have the `Client` accept an executor that the user guarantees will be long-lived (such as the tokio runtime), or maintain a thread ourselves for this purpose.

Reply:
Oops, sorry, @seanmonstar and I talked about this last night but I neglected to follow up with a comment.

TL;DR: I think this was a design mistake with `block_on`, and I'm [correcting it](https://github.com/rust-lang-nursery/futures-rs/pull/890) by having it spawn tasks onto a global thread pool. Eventually that pool will be configurable as well.

But bottom line: we want you to be able to assume that you can use `cx.spawn` freely, even for long-running tasks.

Reply:
Yea, so I think that we should indeed just use `cx.spawn`, and not hold any executor inside the `Client`.

(The tests might need to be changed to stop using `block_on`. At least, it definitely won't work with the published 0.2 beta.)

Reply:
Ah, that looks like a great way to solve the problem!

I'm a bit hesitant to go about changing all the tests, especially since once Aaron's PR is merged, `block_on` will be the simplest way to write that test code.

I guess I could try to shim the new `block_on` using `ThreadPool`?

Reply:
If you wanted to update the `Cargo.toml` to point at the git repo of futures, to get things working, that's fine!

Reply:
I tried this, but then the types will mismatch with `futures 0.2.0-beta`, which `tokio` depends on. I'll try shimming it instead.

Reply:
You need to use the `[replace]` part, so that all dependencies use the same library also.

Reply:
Yay, that works!